silo-to-silo/contracts/SiloToSilo.sol

```diff
@@ -0,0 +1,182 @@
+import "@openzeppelin/contracts/utils/Base64.sol";
+import "@openzeppelin/contracts/access/AccessControl.sol";
+import "./IEvmErc20.sol";
+
+contract SiloToSilo is AccessControl {
+    using AuroraSdk for NEAR;
+    using AuroraSdk for PromiseCreateArgs;
+    using AuroraSdk for PromiseWithCallback;
+
+    bytes32 public constant CALLBACK_ROLE = keccak256("CALLBACK_ROLE");
+
+    uint64 constant BASE_NEAR_GAS = 10_000_000_000_000;
+    uint64 constant WITHDRAW_NEAR_GAS = 50_000_000_000_000;
+    uint64 constant FT_TRANSFER_CALL_NEAR_GAS = 200_000_000_000_000;
+
+    uint128 constant NEAR_STORAGE_DEPOSIT = 12_500_000_000_000_000_000_000;
+
+    NEAR public near;
+    string siloAccountId;
+
+    //[auroraErc20Token] => tokenAccountIdOnNear
+    mapping(IEvmErc20 => string) registeredTokens;
+
+    //[auroraErc20Token][userAddressOnAurora] => userBalance
+    mapping(IEvmErc20 => mapping(address => uint128)) balance;
+
+    event TokenRegistered(IEvmErc20 token, string nearAccountId);
+
+    constructor(address wnearAddress, string memory _siloAccountId) {
+        near = AuroraSdk.initNear(IERC20_NEAR(wnearAddress));
+        siloAccountId = _siloAccountId;
+
+        _grantRole(CALLBACK_ROLE, AuroraSdk.nearRepresentitiveImplicitAddress(address(this)));
+        _grantRole(DEFAULT_ADMIN_ROLE, msg.sender);
+    }
+
+    // TODO: make it trustless
+    function registerToken(IEvmErc20 token, string memory nearTokenAccountId) public onlyRole(DEFAULT_ADMIN_ROLE) {
+        near.wNEAR.transferFrom(msg.sender, address(this), uint256(NEAR_STORAGE_DEPOSIT));
+        bytes memory args = bytes(
+            string.concat('{"account_id": "', getNearAccountId(), '", "registration_only": true }')
+        );
+
+        PromiseCreateArgs memory callStorageDeposit = near.call(
+            nearTokenAccountId,
+            "storage_deposit",
+            args,
+            NEAR_STORAGE_DEPOSIT,
+            BASE_NEAR_GAS
+        );
+        callStorageDeposit.transact();
+
+        registeredTokens[token] = nearTokenAccountId;
+        emit TokenRegistered(token, nearTokenAccountId);
+    }
+
+    function ftTransferCallToNear(
+        IEvmErc20 token,
+        uint128 amount,
+        string calldata receiverId,
+        string calldata message
+    ) external {
+        string storage tokenAccountId = registeredTokens[token];
+        require(address(token) != address(0), "The token is not registered!");
+
+        token.transferFrom(msg.sender, address(this), amount);
+        // WARNING: The `withdrawToNear` method works asynchronously.
+        // As a result, there is no guarantee that this method will be completed before `initTransfer`.
+        // In case of such an error, the user will be able to call `withdraw` method and get his/her tokens back.
+        // We expect such an error not to happen as long as transactions were executed in one shard.
+        token.withdrawToNear(bytes(getNearAccountId()), amount);
```

Comment:
If we will fail here, than user balance will be = amount, but the tokens will be stuck on aurora contract and it will not be easy to withdraw.

Reply:
It could fail just if the only if the account isn't registered, so that's why `registerToken` is exist